embrace-android-sdk/src/main/java/io/embrace/android/embracesdk/EmbraceImpl.kt

```diff
@@ -0,0 +1,1072 @@
+import io.embrace.android.embracesdk.logging.InternalErrorService
+import io.embrace.android.embracesdk.ndk.NdkService
+import io.embrace.android.embracesdk.network.EmbraceNetworkRequest
+import io.embrace.android.embracesdk.network.logging.NetworkCaptureService
+import io.embrace.android.embracesdk.network.logging.NetworkLoggingService
+import io.embrace.android.embracesdk.payload.PushNotificationBreadcrumb
+import io.embrace.android.embracesdk.payload.TapBreadcrumb.TapBreadcrumbType
+import io.embrace.android.embracesdk.prefs.PreferencesService
+import io.embrace.android.embracesdk.session.id.SessionIdTracker
+import io.embrace.android.embracesdk.session.lifecycle.ActivityTracker
+import io.embrace.android.embracesdk.session.lifecycle.ProcessStateService
+import io.embrace.android.embracesdk.session.orchestrator.SessionOrchestrator
+import io.embrace.android.embracesdk.session.properties.SessionPropertiesService
+import io.embrace.android.embracesdk.telemetry.TelemetryService
+import io.embrace.android.embracesdk.utils.PropertyUtils.sanitizeProperties
+import io.embrace.android.embracesdk.worker.WorkerName
+import io.embrace.android.embracesdk.worker.WorkerThreadModule
+import io.opentelemetry.sdk.logs.export.LogRecordExporter
+import io.opentelemetry.sdk.trace.export.SpanExporter
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.regex.Pattern
+
+/**
+ * Implementation class of the SDK. Embrace.java forms our public API and calls functions in this
+ * class.
+ *
+ * Any non-public APIs or functionality related to the Embrace.java client should ideally be put
+ * here instead.
+ */
+@SuppressLint("EmbracePublicApiPackageRule")
+internal class EmbraceImpl @JvmOverloads constructor(
+    private val moduleInitBootstrapper: ModuleInitBootstrapper = ModuleInitBootstrapper()
+) {
+
+    @JvmField
+    val tracer: EmbraceTracer = moduleInitBootstrapper.openTelemetryModule.embraceTracer
+
+    private val uninitializedSdkInternalInterface by lazy<EmbraceInternalInterface> {
+        UninitializedSdkInternalInterfaceImpl(moduleInitBootstrapper.openTelemetryModule.internalTracer)
+    }
+
+    /**
+     * Whether the Embrace SDK has been started yet.
+     */
+    private val started = AtomicBoolean(false)
+    private val sdkClock = moduleInitBootstrapper.initModule.clock
+    private val logger = moduleInitBootstrapper.initModule.logger
+
+    /**
+     * Custom app ID that overrides the one specified at build time
+     */
+    @Volatile
+    private var customAppId: String? = null
+
+    /**
+     * The application being instrumented by the SDK.
+     */
+    @Volatile
+    var application: Application? = null
+        private set
+
+    /**
+     * The type of application being instrumented by this SDK instance, whether it's directly used by an Android app, or used via a hosted
+     * SDK like Flutter, React Native, or Unity.
+     */
+    @Volatile
+    private var appFramework: Embrace.AppFramework? = null
```

Comment:
In a subsequent change, I want to get rid of these potentially nullable things with a default implementation used before the SDK initializes. These are a scourge